assets/js/base/hooks/customer/use-customer-data.js

```diff
@@ -19,73 +20,94 @@ export const useCustomerData = () => {
 	} = useStoreCart();
 
 	const [ customerData, setCustomerData ] = useState( {
-		billingData: initialBillingData,
-		shippingAddress: initialShippingAddress,
+		billingData: cartBillingData,
+		shippingAddress: cartShippingAddress,
 	} );
-	const previousCustomerData = useRef( customerData );
-	const [ debouncedCustomerData ] = useDebounce( customerData, 400 );
 
+	const currentBillingData = useRef( cartBillingData );
+	const currentShippingAddress = useRef( cartShippingAddress );
+	const [ debouncedCustomerData ] = useDebounce( customerData, 400, {
+		equalityFn: ( prevData, newData ) => {
+			return ! (
+				isShallowEqual( prevData.billingData, newData.billingData ) ||
+				isShallowEqual(
+					prevData.shippingAddress,
+					newData.shippingAddress
+				)
+			);
+		},
+	} );
+
+	/**
+	 * Set billing data.
+	 *
+	 * Contains special handling for email and phone so those fields are not overwritten if simply updating address.
+	 */
 	const setBillingData = useCallback( ( newData ) => {
-		setCustomerData( ( prevState ) => ( {
-			...prevState,
-			billingData: {
-				...prevState.billingData,
-				...newData,
-			},
-		} ) );
+		setCustomerData( ( prevState ) => {
+			return {
+				...prevState,
+				billingData: {
+					...prevState.billingData,
+					...newData,
+				},
+			};
+		} );
 	}, [] );
 
 	const setShippingAddress = useCallback( ( newData ) => {
 		setCustomerData( ( prevState ) => ( {
 			...prevState,
-			shippingAddress: {
-				...prevState.shippingAddress,
-				...newData,
-			},
+			shippingAddress: newData,
 		} ) );
 	}, [] );
 
-	// This will push both billing and shipping addresses to the server as a single request.
 	useEffect( () => {
-		const shouldUpdateBilling = shouldUpdateAddressStore(
-			previousCustomerData.current.billingData,
-			debouncedCustomerData.billingData
-		);
-		const shouldUpdateShipping = shouldUpdateAddressStore(
-			previousCustomerData.current.shippingAddress,
-			debouncedCustomerData.shippingAddress
-		);
-		if ( shouldUpdateBilling || shouldUpdateShipping ) {
-			// Keep refs in sync to avoid multiple calls to API if debounce triggers again.
-			previousCustomerData.current = debouncedCustomerData;
-
-			const requestData = {};
-
-			if ( shouldUpdateBilling ) {
-				requestData.billing_address = {
-					...debouncedCustomerData.billingData,
-					phone: debouncedCustomerData.billingData.phone || null,
-					email: debouncedCustomerData.billingData.email || null,
-				};
-			}
+		if ( ! isShallowEqual( currentBillingData.current, cartBillingData ) ) {
+			currentBillingData.current = cartBillingData;
+		}
+	}, [ cartBillingData ] );
 
-			if ( shouldUpdateShipping ) {
-				requestData.shipping_address =
-					debouncedCustomerData.shippingAddress;
-			}
+	useEffect( () => {
+		if (
+			! isShallowEqual(
+				currentShippingAddress.current,
+				cartShippingAddress
+			)
+		) {
+			currentShippingAddress.current = cartShippingAddress;
+		}
+	}, [ cartShippingAddress ] );
```

Comment:
Could the `currentShippingAddress` and `currentBillingAddress` still be in a single ref object and merge these two above effects so you'd end up with something like this?

```js
useEffect( () => {
  if (
    ! isShallowEqual(
      currentCustomerData.current.shippingAddress,
      cartShippingAddress
    )
  ) {
    currentCustomerData.current.shippingAddress = cartShippingAddress;
  }
  if (
    ! isShallowEqual(
      currentCustomerData.current.billingData,
      cartBillingData
    )
  ) {
    currentCustomerData.current.billingData = cartBillingData;
  }
}, [ cartShippingAddress, cartBillingData ] );
```

Reply:
I'm going to apply this feedback, however, I'm not sure what is really best practice or more performant. The above useEffect will be triggered more often than just the billing or shipping one, so maybe `isShallowEqual` will run more often?